cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java

```diff
@@ -103,6 +104,13 @@ protected void shutDown() {
   }
 
   /**
+   * Returns the messaging context.
+   */
+  protected MessagingContext getMessagingContext() {
+    return messagingContext;
```

Comment:
You need to pass in the messaging context to the `MultiThreadDatasetCache` created above so that it can participant in transaction, which is needed for the publishing of program state that you added in this PR.